module/VuFind/src/VuFind/I18n/Initializer.php

```diff
@@ -0,0 +1,289 @@
+use Zend\Mvc\I18n\Translator as MvcTranslator;
+use Zend\View\Model\ViewModel;
+
+/**
+ * Handles I18n initialization.
+ *
+ * @category VuFind
+ * @package  Translator
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Sebastian Kehr <kehr@ub.uni-leipzig.de>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Site
+ */
+class Initializer
+{
+    /**
+     * Translator.
+     *
+     * @var I18nTranslator
+     */
+    protected $translator;
+
+    /**
+     * Language base directories.
+     *
+     * @var string[]
+     */
+    protected $baseDirs = [];
+
+    /**
+     * File loader.
+     *
+     * @var ExtendedIni
+     */
+    protected $loader;
+
+    /**
+     * Initializer constructor.
+     *
+     * @param Request          $request       Request
+     * @param Config           $config        Configuration
+     * @param CookieManager    $cookies       Cookie manager
+     * @param StorageInterface $cache         Language cache
+     * @param MvcTranslator    $mvcTranslator Translator
+     * @param ViewModel        $viewModel     ViewModel
+     */
+    public function __construct(
+        Request $request,
+        Config $config,
+        CookieManager $cookies,
+        StorageInterface $cache,
+        MvcTranslator $mvcTranslator,
+        ViewModel $viewModel
+    ) {
+        $language = $this->detectLanguage($config, $request, $cookies);
+
+        $this->translator = $mvcTranslator->getTranslator();
+        $this->loader = $this->translator->getPluginManager()
+            ->get(ExtendedIni::class);
+
+        $this->translator->setLocale($language);
+        $this->translator->setCache($cache);
+
+        $this->loader->setFallbacks($this->parseFallbacks($config));
+
+        $rtlLangs = isset($config->LanguageSettings->rtl_langs)
+            ? array_map(
+                'trim', explode(',', $config->LanguageSettings->rtl_langs)
+            ) : [];
+
+        $viewModel->setVariable('userLang', $language);
+        $viewModel->setVariable('allLangs', $config->Languages);
+        $viewModel->setVariable('rtl', in_array($language, $rtlLangs));
+
+        $this->addBaseDir(APPLICATION_PATH . '/languages');
+        $this->addBaseDir(LOCAL_OVERRIDE_DIR . '/languages');
+    }
+
+    /**
+     * Initializes the translator and loader.
+     *
+     * @return void
+     */
+    public function init()
+    {
+        $cache = $this->translator->getCache();
+        $cacheKey = md5(implode('', $this->baseDirs));
+
+        if (!$cache->hasItem($cacheKey)) {
+            $cache->setItem($cacheKey, $this->detectTextDomains());
+        }
+
+        foreach ($cache->getItem($cacheKey) as $textDomain) {
+            $this->translator->addRemoteTranslations(
+                ExtendedIni::class, $textDomain
+            );
+        }
+
+        $this->loader->setDirs(array_reverse($this->baseDirs));
+    }
```

Comment:
Pass the base directories to the loader in reversed order so most recently added directories will have top most priority.

Reply:
Some of your comments (like this one) would probably be useful as comments in the code itself, for future reference.